MM-42810: Introduce a channel hook for a websocket event

agnivade opened the pull request:
Sometimes a broad distinction of just a channelID or a userID
is not enough to efficiently send a websocket event to users.

In several cases, depending on the user and channel, we might
need to modify the message. Therefore, we introduce the
concept of a channel hook that will get executed if the scope
is set to a channel. This hook can be populated at the app layer
to perform any application specific logic to the event.

Care must be taken to avoid race conditions as the passed event
is not deep copied. It is left to the user to treat it carefully.

For this issue, the main problem was that since we don't know
which users have permissions to which channels, we had to go through
_all_ members of a channel to figure that out. This was redundant
since a large portion of those users might not even be connected
at that time.

We solve this with the channel hook where we push this check
to be performed later while actually sending the event. This
reduces the computation to be done only for _connected_ users
rather than _all_ users of a channel.

The next iteration of this should be to use websocket subscriptions
to monitor exactly which users are on that channel to even
trim down that list. That is a larger initiative to be taken later.

Tested locally with a channel of 50 users. Here are rough results:
```
With PR:
patchPost 97ms
createPost 90ms

Master:
patchPost 306ms
createPost - 298ms
```

https://mattermost.atlassian.net/browse/MM-42810

```release-note
Improve  performance while sending messages
with permalinks to channels with large number of users.
```

lieut-data commented on server/channels/app/post.go:
```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	broadcastCopy := messageCopy.GetBroadcast()
+	broadcastCopy.ChannelHook = func(userID string, ev *model.WebSocketEvent) bool {
+		if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, userID, permalinkPreviewedChannel) {
+			postCopy.Metadata.Embeds[0].Data = nil
```
Is `postCopy` the same instance each time we call this hook? At first glance, it seems like the first user without permission to the channel will modify the shared `postCopy` for all subsequent users whether or not they have access.

agnivade replied:
In my testing, that didn't happen. However, I think I might have been just lucky. I will verify this deeply. It looks like I need to clone inside the hook.

lieut-data reviewed (changes requested):
Wow, the potential impact of this change is tremendous! A few clarifying thoughts as I'm wondering if the behaviour is changing unintentionally.

Also, two follow up questions:
* Is there an opportunity to expand unit test coverage here?
* What's the expected impact in ~~a) queuing N websocket events when we only queued 1 before,~~ (nvm, I typed this up too quickly!) and 2) adding a closure to each of those events? Will `createPost` get faster at the expense of more memory pressure?

lieut-data commented:
> For this issue, the main problem was that since we don't know
which users have permissions to which channels, we had to go through
all members of a channel to figure that out. This was redundant
since a large portion of those users might not even be connected
at that time.

Another question: if most of the channel members are connected, do we still glean any savings? The issue in question primarily arose for me from permalink posts in a channel like town square, and I'm wondering if even having 30-40% of the members connected means we'll still have the fundamental issue at play here.

agnivade commented:
> Another question: if most of the channel members are connected, do we still glean any savings? 

We don't. And that's where the last line of my PR comes into play ;)

> The next iteration of this should be to use websocket subscriptions
to monitor exactly which users are on that channel to even
trim down that list. That is a larger initiative to be taken later.

lieut-data commented:
> The next iteration of this should be to use websocket subscriptions
to monitor exactly which users are on that channel to even
trim down that list.

Would the idea be to not send these payloads if the users are simply looking at a different channel? Would that case a thundering herd of users switch to an `@here` in `~town-square` and trying to resolve the permalink preview via an API request?

agnivade commented on server/channels/app/platform/web_conn.go:
```diff
@@ -789,6 +800,13 @@ func (wc *WebConn) ShouldSendEvent(msg *model.WebSocketEvent) bool {
+		if msg.GetBroadcast().ChannelHook != nil && msg.GetBroadcast().ChannelHook(wc.UserId, msg) {
+			// If hook returns true, that means message has been modified. We need
+			// to wipe off the pre-computed JSON
+			msg.RemovePrecomputedJSON()
```
Nah, it just falls back to marshalling the JSON. It's a tradeoff. SInce we are modifying the message content in the hook, the existing precomputed JSON is obsolete. Therefore we need to marshal it from scratch again.

agnivade commented on server/channels/app/post.go:
```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
+	}
+	// Using DeepCopy here to avoid a race condition
+	// between publishing the event and setting the "post" data value below.
+	messageCopy := message.DeepCopy()
+	broadcastCopy := messageCopy.GetBroadcast()
```
Hmm yeah .. I will check this.

agnivade commented on server/channels/app/post.go:
```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	messageCopy := message.DeepCopy()
+	broadcastCopy := messageCopy.GetBroadcast()
+	broadcastCopy.ChannelHook = func(userID string, ev *model.WebSocketEvent) bool {
+		if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, userID, permalinkPreviewedChannel) {
```
Ah indeed that was the only usage for that method! Yep, will clean up :)

agnivade commented on server/channels/app/post.go:
```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+		if err != nil {
+			mlog.Warn("Failed to encode post to JSON", mlog.Err(err))
+		}
+		ev.AddWithCopy("post", postJSON)
```
> and we never set post on messageCopy to begin with.

We do. In the higher layers.

lieut-data replied:
> We do. In the higher layers.

Eek. So previously we unambiguously set the `post` property from within this method before publishing, but now it's required that the caller do this otherwise this method won't work correctly for some users.

agnivade replied:
I think the expectation for this method is that - if the post contains a permalink, then it will do whatever is necessary to make changes to the websocket event and publish it. But if it does not, then publishing falls on the higher layer itself. With that logic, I think it makes sense that the caller already set the `post` property before calling. Because if there's no permalink, the caller has to publish the post.

lieut-data replied:
> I think the expectation for this method is that - if the post contains a permalink, then it will do whatever is necessary to make changes to the websocket event and publish it. But if it does not, then publishing falls on the higher layer itself. With that logic, I think it makes sense that the caller already set the post property before calling. Because if there's no permalink, the caller has to publish the post.

I think the coupling is still wrong -- here's a (modified) implementation that matches your expectations of the caller but would be broken in the case I described:
```go
	published, err := a.publishWebsocketEventForPermalinkPost(c, rpost, message)
	if err != nil {
		return nil, err
	}
	if !published {
		message.Add("post", postJSON)
		a.Publish(message)
	}

lieut-data commented on server/channels/app/post.go:
```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
-		postJSON, jsonErr := postForUser.ToJSON()
-		if jsonErr != nil {
-			mlog.Warn("Failed to encode post to JSON", mlog.Err(jsonErr))
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
+	}
```
> Then posting this shouldn't expand that post for them.

Oh, got it: so the caller of `createPost` is getting this post object back, and since we've set `post.Metadata.Embeds[0].Data` above, we need to unset it here after making a copy. But that means in the happy case when the user does have access, this method also has the surprising side effect of resolving `post.Metadata.Embeds[0].Data` via a method called `publishWebsocketEventForPermalinkPost`.

Looking at the old code, it's not obvious "where" this used to happen. Does it stand out to you?

agnivade replied:
It used to happen in `sanitizePostMetadataForUserAndChannel`. I agree that this has become slightly confusing. I'll look into this separately in a further PR. Thanks for raising this.

lieut-data commented on server/channels/app/post.go:
```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	} else {
+		return false, nil
```
I think the various discussions below highlight some of the surprising behaviour, but to summarize, this method has implicit contracts with the caller (be sure to `.Add("post", postJSON)`) and undocumented side effects (populating `post.Metadata.Embeds[0].Data` if the user has access to it).

I wonder if most of this could just be resolved with a `publishWebsocketEventForPost` that encapsulates both the logic in this method and the fiddly bits that callers seem to have to juggle, e.g. all of this:
```go
	message.Add("post", postJSON)

	published, err := a.publishWebsocketEventForPermalinkPost(c, rpost, message)
	if err != nil {
		return nil, err
	}
	if !published {
		a.Publish(message)
	}
```

I'm 0/5 on whether we do this in this PR: just thinking out loud after reading this (existing) code for the first time in a while. I also want to balance the new complexity we're adding with the existing potential for confusion.

agnivade replied:
Indeed, I agree. I will think some more on this, but I'd like to keep that refactor as a separate PR and focus this PR purely on the performance improvement.

agnivade commented:
@lieut-data - PTAL

 - Removed unnecessary messageCopy
  - Reworked sanitizePostMetadataForUserAndChannel to be used from the hook
  so that unit test coverage exists.
  - Also modified it to return a true/false rather than actually modify the post.
  This simplifies the data race issues while handling the post metadata. Since
  we just have to deal with metadata on/off, we cache the modified JSON with no metadata
  and just check whether the user has access or not. If they don't, then we attach the
  modified JSON. This also avoids the need to JSON marshal every time.
  - Added more comments to explain what's going on.

streamer45 commented on server/channels/app/platform/web_conn.go:
```diff
@@ -789,6 +800,13 @@ func (wc *WebConn) ShouldSendEvent(msg *model.WebSocketEvent) bool {
+		if msg.GetBroadcast().ChannelHook != nil && msg.GetBroadcast().ChannelHook(wc.UserId, msg) {
+			// If hook returns true, that means message has been modified. We need
+			// to wipe off the pre-computed JSON
+			msg.RemovePrecomputedJSON()
```
Why don't we leave this burden to the hook definition? On that side we have access to the message and the necessary context to make the decision right?

agnivade replied:
That's why I am returning the true/false. It didn't look to me something that the hook should be aware about. The app developer shouldn't be aware about precomputed JSON. That's why I chose to do it in the web_conn layer.

streamer45 commented on server/channels/app/post.go:
```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
```
Is there any point in leaving this case?

agnivade replied:
Nope, as Jesse pointed out below. The latest refactor made this redundant. I'll remove.

lieut-data commented on server/channels/app/post.go:
```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	// multiple goroutines would be racing to modify the Metadata field.
+	// And deep-copying the Embed data also has challenges because the type is "any".
+	post.Metadata.Embeds[0].Data = nil
+	modifiedJSON, err := post.ToJSON()
```
Would it be clearer if we called this `postWithoutPermalinkPreviewJSON`?

lieut-data commented on server/channels/app/post.go:
```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	// This simplifies the data handling in the channel hook, where otherwise
+	// multiple goroutines would be racing to modify the Metadata field.
+	// And deep-copying the Embed data also has challenges because the type is "any".
+	post.Metadata.Embeds[0].Data = nil
```
Obviously we do this now because we set it above, but wondering if -- given this method /owns/ setting this embed -- we can just skip it altogether after simplifying above?

agnivade replied:
Yep good point.

streamer45 commented on server/public/model/websocket_message.go:
```diff
@@ -189,10 +190,21 @@ func (ev *WebSocketEvent) PrecomputeJSON() *WebSocketEvent {
 	return copy
 }
 
+func (ev *WebSocketEvent) RemovePrecomputedJSON() {
+	ev.precomputedJSON = nil
+}
+
 func (ev *WebSocketEvent) Add(key string, value any) {
 	ev.data[key] = value
 }
 
+// AddWithCopy copies the map and writes to a copy of that,
+// and sets the new map to the event
+func (ev *WebSocketEvent) AddWithCopy(key string, value any) {
+	ev.data = copyMap(ev.data)
+	ev.data[key] = value
+}
```
Are these safe? From what I remember we used to avoid mutating ws events at all costs and always returned a copy when having to change something.

agnivade replied:
Hmm .. good point, let me change that.

agnivade replied:
Ah no, it has further problems. The issue is that we are calling this via `ShouldSendEvent` from the web_hub, but just passing the pointer. This means to modify that, we either need to assign the return value or do something like `*dst = *src`. Both seem equally invasive to me.

I tested the current code extensively in `-race` mode and didn't see any issue. I am inclined to keep it this way unless you have strong opinions?

streamer45 replied:
Okay, noting that this is safe only because we call [`PrecomputeJSON()`](https://github.com/mattermost/mattermost/blob/cc2205c6a486fdd3dd42e3a8de5a329e2ca685c4/server/channels/app/platform/web_hub.go#L489) which shallow copies the message, including the map reference, before the hook runs. Same goes for calling `RemovePrecomputedJSON()`. So extra care should be taken when using this method because it's not inherently safe.

agnivade replied:
Yep correct.

lieut-data commented on server/channels/app/post.go:
```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	// This wipes off the original post content. And then the marshalled post content
+	// in the websocket event is taken care of in the channel hook.
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
```
I'm still confused here: we set above, then unset, then set again, and now we unset if the posting user doesn't have permission. Where do we set it for users that /should/ get it?

agnivade replied:
Via the websocket event :) That's what I tried to clarify in the comment. But it's hard to grasp.

Essentially, one object is returned from the REST API endpoint, the other objects are sent via websocket. L818 is about the object that gets returned from the REST endpoint. For everyone else (including the posting user), it gets taken care of by the websocket event.

lieut-data replied:
I wanted to test this behaviour to see what's actually happening. I made a private channel in which the left user was a member, but the right user was not. On master, either user posting a permalink to a post in that channel correctly displays for the left user and not for the right user. 

Then I checked out this PR's branch, refreshed both clients, and posted the same permalink again for each user. Now, neither user sees the permalink resolve, even though the left user should have access to the channel.

Can you reproduce this locally, @agnivade?

![CleanShot 2023-06-23 at 10 00 46@2x](https://github.com/mattermost/mattermost/assets/1023171/f1170ae8-876c-44ca-8143-050971068f63)

lieut-data commented on server/channels/app/post.go:
```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+		return true
 	}
+	message.SetBroadcast(broadcastCopy)
+	a.Publish(message)
```
I once again spent too long wondering where the `"post"` was attached before remembering it was on the caller's side. </datapoint>

lieut-data commented on server/channels/app/post.go:
```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	broadcastCopy.ChannelHook = func(userID string, ev *model.WebSocketEvent) bool {
+		hasChange := a.shouldSanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedChannel, userID)
+		if !hasChange {
+			// If there is no change, then the original post which was attached
+			// will get sent.
+			return false
 		}
-		messageCopy.Add("post", postJSON)
-		a.Publish(messageCopy)
+
+		ev.AddWithCopy("post", modifiedJSON)
```
This block reads "backwards" from how we usually do "security things": default to denying, and only if we know it's safe, return the sensitive information. Would it make sense to invert the logic to cache the `postWithPermalinkJSON` and only set it if `!shouldSanitize`?

agnivade replied:
The issue is that the `postWithPermalinkJSON` is already set from the higher layers and out of control for this method.

streamer45 reviewed (commented):
Nice! Left some comments, plus a couple of extra questions:

1. Should we write a test?
2. Would this benefit from a quick pass from security?

lieut-data commented on server/channels/app/post.go:
```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
+	broadcastCopy.ChannelHook = func(userID string, ev *model.WebSocketEvent) bool {
+		hasChange := a.shouldSanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedChannel, userID)
+		if !hasChange {
+			// If there is no change, then the original post which was attached
+			// will get sent.
```
I stumbled over this again, thinking the "original post" was the post set by the caller, and wondering where the permalink preview got set, before remembering that we (sometimes) set it above, so it's not so "original" anymore as the comment suggests. More and more, I'm wondering if we should bite the bullet and just fix this with a proper `publishWebsocketEventForPost`. It's extremely hard to reason about the correctness of the existing code + changes.

agnivade replied:
The only thing that I am wondering is that, in that case we'll be doing `ev.Add()` inside the method but there'll be a lot of other `ev.Add()` outside the method as well for other stuff. For readers going through the higher layer, they'll wonder "where is the `post` attribute getting set?" only to find out there's this one method which sets it.

lieut-data commented on server/channels/app/post_metadata.go:
```diff
@@ -190,16 +190,19 @@ func (a *App) getEmbedsAndImages(c request.CTX, post *model.Post, isNewPost bool
+// shouldSanitizePostMetadataForUserAndChannel returns true/false whether the user
+// has permission to view the metadata, and thereby the metadata should be
+// wiped off or not.
+func (a *App) shouldSanitizePostMetadataForUserAndChannel(c request.CTX, post *model.Post, previewedChannel *model.Channel, userID string) bool {
```
FWIW, I'm still in favour of inlining this. It's far too specific to the permalink preview case to be useful in any other context, and it would actually eliminate more code since some of these checks are redundant.

agnivade replied:
Yeah, the thing is then this becomes hard to unit test. Let me see.

lieut-data reviewed (changes requested):
Thanks, @agnivade. Apologies, but I'm still stumbling over something below -- see in particular the question about the post author clearing the permalink preview for other users. It's possible I'm just stumbling over the complexity here, but -- as noted below -- I'm inclined to just fix this holistically with a `publishWebsocketEventForPost` that wouldn't have to jump through the hoops below (e.g. "hack to cache").

hmhealey commented:
I'm even less than 0/5 on this since I'm proposing an entirely different way of supporting this case, but what if, instead of sanitizing the websocket event for certain users, we instead extend the `broadcast` parameter to be allow for more control over the event and then we send a sanitized and unsanitized one slightly different `broadcast` settings? Like for this case, we could prepare two messages: 

1. The first one would have the post attached and be sent to users who are both in the post's channels and the permalinked channel.
2. The second one wouldn't have the post attached and would be send to users who are in the post's channel but not the permalinked channel.

That obviously requires extra work to support logical rules, and we'd have to make sure to maintain compatibility with clients that rely on the `broadcast` option in its current state, but that feels like it would be easier to reason about compared to having to understand when the event gets sanitized.

Feel free to ignore this though since I can't stop myself from trying to find different ways to do things that already have good solution 😛

agnivade commented:
@streamer45 @lieut-data - LMK how does this new iteration look:

- Made the check inline
- Invert the check in the hook to be more security friendly
- Changing publish event to do everything
- Added test for HasPermissionToReadChannel
- Added test for the channel hook

agnivade commented:
>  Like for this case, we could prepare two messages:
The first one would have the post attached and be sent to users who are both in the post's channels and the permalinked channel.
The second one wouldn't have the post attached and would be send to users who are in the post's channel but not the permalinked channel.

@hmhealey - Can you expand further on the implementation of this? How will you find out which connected users have permission to the permalinked channel _before_ entering the websocket layer? And if you don't then you get into the same problem.

Therefore to know which users are connected, you need to enter the websocket layer, and then you will need some sort of hook at that point to check if the user has the permission or not.

Let me know.

hmhealey commented:
I'm sort of meaning that we could eventually change the Broadcast field to more or less give it some basic boolean logic for things like filtering channel IDs. Sketching something out, it might look like
```go
func SendCreatePostEvents(post *model.Post, permalinkedPost *model.Post) {
	if permalinkedPost != nil {
		send(&model.WebSocketMessage{
			Type: model.WebsocketEventPosted,
			Broadcast: {
				Filter: And(
					MemberOfChannel(post.ChannelId),
					NotMemberOfChannel(permalinkedPost.ChannelId),
				),
			},
			Data: {
				Post: post,
			},
		})

		send(&model.WebSocketMessage{
			Type: model.WebsocketEventPosted,
			Broadcast: {
				Filter: And(
					MemberOfChannel(post.ChannelId),
					MemberOfChannel(permalinkedPost.ChannelId),
				),
			},
			Data: {
				Post: post.WithEmbeddedPostMetadata(permalinkedPost),
			},
		})
	} else {
		send(&model.WebSocketMessage{
			Type: model.WebsocketEventPosted,
			Broadcast: {
				Filter: MemberOfChannel(post.ChannelId),
			},
			Data: {
				Post: post,
			},
		})
	}
}
```

We'd still have the WebSocket code itself handling permissions checking, and it could do it in a smarter way like you're doing in this PR, but I think this could be a nice future change to make it very clear to someone reading the code who should receive which message. As it is, we're kind of limited by the structure of `Broadcast` which is what makes these cases hard like with sanitizing for non-admins or having more specific messages for posts with permalink previews.

But like I said, that's entirely a different direction than any of our existing code, and it would require rewriting a ton of stuff, so I was pitching it as a totally hypothetical thing

agnivade commented:
Gentle ping @lieut-data @streamer45

lieut-data reviewed (approved):
Thanks for the iteration again, @agnivade! A few final thoughts below that become more apparent with the clarity your changes have brought to the original implementation -- but not blocking.

agnivade commented on server/channels/app/post.go:
```diff
@@ -741,85 +740,94 @@ func (a *App) UpdatePost(c *request.Context, post *model.Post, safeUpdate bool)
-	for _, cm := range channelMembers {
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
+	if permalinkPreviewedPost == nil {
+		return nil
+	}
```
Yep, good point. Let me do that.

agnivade commented on server/channels/app/post.go:
```diff
@@ -741,85 +740,94 @@ func (a *App) UpdatePost(c *request.Context, post *model.Post, safeUpdate bool)
-		broadcastCopy := messageCopy.GetBroadcast()
-		broadcastCopy.UserId = cm.UserId
-		messageCopy.SetBroadcast(broadcastCopy)
+	// In case the user does have permission to read, we set the metadata back.
```
done

agnivade commented:
@lieut-data - Would you mind giving another look at the latest commit? I moved around some code. I asked for a review again just to be safe.

streamer45 reviewed (commented):
Looks good! 

How hard would it be to provide a test that covers this new functionality? Mostly checking that the hook gets executed plus the behaviour in case of positive return (i.e. message content has changed).

agnivade commented on server/public/model/websocket_message.go:
```diff
@@ -92,14 +92,15 @@ type WebSocketMessage interface {
+	UserId                string                                       `json:"user_id"`                           // broadcast only occurs for this user
+	OmitUsers             map[string]bool                              `json:"omit_users"`                        // broadcast is omitted for users listed here
+	ChannelId             string                                       `json:"channel_id"`                        // broadcast only occurs for users in this channel
+	ChannelHook           func(userID string, ev *WebSocketEvent) bool `json:"-"`                                 // ChannelHook is a function that runs for a channel scoped event. It can be used to modify the event payload based on some custom logic that runs only for connected users.
```
Sure

lieut-data commented on server/channels/app/post.go:
```diff
@@ -766,22 +766,27 @@ func (a *App) publishWebsocketEventForPost(c request.CTX, post *model.Post, mess
+		return model.NewAppError("publishWebsocketEventForPost", "app.post.marshal.app_error", nil, "", http.StatusInternalServerError).Wrap(jsonErr)
+	}
+
+	previewedPost, appErr := a.GetSinglePost(permalinkPreviewedPost.PostID, false)
```
More thoughts -- unrelated to this PR, but while we're here! -- why do we have to `a.GetSinglePost` and `a.GetChannel` at all? The `PreviewPost` contains all the information we need to do the permission checks. So it just comes down to whether or not we trust the caller not to provide bogus data?

agnivade replied:
We can do away with `GetSinglePost`, but we need `GetChannel` to get the channel.TeamId for the `a.HasPermissionToReadChannel` check.

lieut-data reviewed (approved):
Looks good to me, @agnivade. But I couldn't help observe one more thing for a potential future PR ;P

streamer45 commented on server/channels/app/platform/web_hub_test.go:
```diff
@@ -126,6 +127,81 @@ func TestHubStopRaceCondition(t *testing.T) {
+	// Wait until the goroutines from NewWebConn are finished.
+	th.Service.waitForGoroutines()
+	th.TearDown()
+	assert.Equal(t, true, test.Load())
```
We could use `assert.True` I suppose but clearly not a big deal.

streamer45 commented on server/channels/app/platform/web_hub_test.go:
```diff
@@ -126,6 +127,81 @@ func TestHubStopRaceCondition(t *testing.T) {
+	ev := model.NewWebSocketEvent("", "", "channelID", "", nil, "")
+	broadcast := ev.GetBroadcast()
+	var test atomic.Bool
+	broadcast.ChannelHook = func(_ string, ev *model.WebSocketEvent) bool {
+		test.Store(true)
+		return true
+	}
```
Could we add a simple check to test that the precomputed JSON has been reset?

agnivade replied:
It's a bit difficult from the app layer because the field isn't exported. Do you see a simple way to check this?

streamer45 replied:
Something like `reflect.ValueOf(ev).Elem().FieldByName("precomputedJSON")` should do the trick.

agnivade replied:
I hope you are not serious :joy:

streamer45 replied:
Just being pragmatic :) Are we reluctant to use `reflect` in a test? Otherwise you likely need to build an interface and mock it. It seems like a lot of pain. 

Of course we could leave it as is, but it seemed important to check that the cached json was dealt with, mostly for security reasons.

streamer45 replied:
> Using reflect to access private fields like that is a sure sign of poorly written code.

That's concerning :) I don't think we would normally ever need to access it though. This is about testing across packages without bloating with mocks.

> But we can give this a security review to be on the safe side. Would that alleviate your concerns?

Sure a security review can't hurt.

streamer45 reviewed (approved):
Thanks @agnivade, left a couple of minor suggestions on the test but nothing blocking.

mattermost-build commented:
You don't have permissions to trigger this command.
 It's only available for organization members.

agnivade commented:
Hi @DHaussermann - this needs some testing around permalink previews. Steps are mentioned in the JIRA ticket. Let me know if you have any questions.  Thanks!

agnivade commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/348440) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/348440)

agnivade commented:
@jupenur - Requesting a quick security review on this one.

agnivade commented:
@DHaussermann could you give this a test whenever you get a chance? Thanks!

agnivade commented:
Juho seems to be on PTO. @esarafianou could you give this a look?

jupenur commented on server/channels/app/post.go:
```diff
@@ -741,85 +740,89 @@ func (a *App) UpdatePost(c *request.Context, post *model.Post, safeUpdate bool)
-		if err.StatusCode == http.StatusNotFound {
-			mlog.Warn("channel containing permalinked post not found", mlog.String("referenced_channel_id", previewedPost.ChannelId))
-			return false, nil
+	previewedPost, appErr := a.GetSinglePost(permalinkPreviewedPost.PostID, false)
```
There is a race here. If the permalinked post gets deleted after entering `publishWebsocketEventForPost` but before reaching this line, `GetSinglePost` here will return an error, `publishWebsocketEventForPost`, and the deferred `Publish` runs without setting the `ChannelHook`. Which means the deleted post's contents get broadcast to all users in the channel unsanitized.

agnivade replied:
> If the permalinked post gets deleted after entering publishWebsocketEventForPost but before reaching this line

The `createPost` endpoint doesn't return until it returns from `publishWebsocketEventForPost` and finishes the other stuff. Were you thinking of a case where the DB is manually modified?

Would it be safer for you if we move the `message.Add("post", postJSON)` to before L804? Just to prevent sending a post with full metadata?

jupenur replied:
Why does it matter when the `createPost` endpoint returns? This is the case I'm thinking about:

1. User `A` creates post `a` in a private channel
2. User `B` creates post `b` in a public channel and links to post `a`
3. Simultaneously with step `2`, user `A` deletes post `a`
4. All users in the public channel receive a WS broadcast about the creation of post `b`, containing the embedded post `b`

`createPost` returns some time after step `4` but at that point the WS broadcast has already gone out.

agnivade replied:
Ah yes, sorry I had a brain fart. So we have two options: 
- Either don't publish in case there's an error
- Move the `message.Add` line below these checks.

I am leaning towards option 1 as that looks more in line with the earlier implementation.

jupenur replied:
A third option would be to set the `ChannelHook` early. The variables it uses, like `permalinkPreviewedChannel`, would have to be initialized with empty values, but having them empty when the function returns early is a safe failure mode.

I'm ok with any of the three options.

jupenur commented on server/channels/app/post.go:
```diff
@@ -742,85 +741,95 @@ func (a *App) UpdatePost(c *request.Context, post *model.Post, safeUpdate bool)
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	var permalinkPreviewedChannel *model.Channel
+	permalinkPreviewedChannel, appErr = a.GetChannel(c, previewedPost.ChannelId)
+	if appErr != nil {
+		if appErr.StatusCode == http.StatusNotFound {
+			mlog.Warn("channel containing permalinked post not found", mlog.String("referenced_channel_id", previewedPost.ChannelId))
+			return nil
```
There's still a race here, albeit one that's unlikely to ever materialize in practice: What if the channel gets permanently deleted (not archived) between the calls `GetSinglePost` and `GetChannel`? Since not finding the channel is not an error, `Publish` gets run again without sanitizing the metadata.

agnivade replied:
That is the existing state as of today, so I think we should treat that separately if we want to.

jupenur replied:
No, this is what the equivalent bit is in the current code: https://github.com/mattermost/mattermost/blob/cc6de45bdeb2f3dbf889bd81e51c91a38d87d70d/server/channels/app/post.go#L791-L798

Notice how the `published` return value is `false` even when there is no error. The current code doesn't use `defer` for `Publish`.

agnivade replied:
It doesn't use defer, but the result is the same if I am reading the code right:

```go
published, err := a.publishWebsocketEventForPermalinkPost(c, rpost, message)
	if err != nil {
		return nil, err
	}
	if !published {
		a.Publish(message)
	}
```

jupenur replied:
Well crap, you're right. Yeah let's deal with this separately. I'll create a ticket.

DHaussermann reviewed (approved):
Tested and passed
- Regression tested business logic
  - Preview will show if the the  post is in public channel regardless of membership
  - Preview will not show if the the post is in private channel and test user does not not have membership
  - Preview will not show if the the test user does not not have team membership
- in the webapp creating a post with a permalink saw no noticeable delay even in a channel with 1000 users

LGTM!

Thanks for the help and steps on this @agnivade
